### ISSUE TITLE
[Question] How do I extract data from this app?

### ISSUE THREAD

**Issue by Davetheraveuk** [dglent/meteo-qt#137]
Hi, I would like to extract data from this app and display it in conky. How do I do this? Thanks, Dave.

**Comment by dglent** [comment]
Hello,

Do you know how to add widgets in Conky ?
Maybe the log file in ~/.config/meteo-qt/meteo-qt.log can help (and add some more specific information for Conky?)

**Comment by Davetheraveuk** [comment]
@dglent 
Hi, yes, I am fully aware of conky and its widgets (see pic attacted).
I had a look at the log file and this is all that was in there:

`
2023/09/11 19:11:50 ERROR: Your openweathermap key has no access to onecall api - 2781: meteo_qt

2023/09/11 19:11:51 INFO: Actual weather status for: Loughton GB 21.9°

Feels like 22 °C

overcast clouds - 2431: meteo_qt
`

I have no wish to see any meteo-qt icons within conky, just the data from the meteo-gt app.
Hope you can help
Thanks.
Dave.

![Screenshot from 2023-09-11 19-16-23](https://github.com/dglent/meteo-qt/assets/11774054/191dbbfb-5e86-4d06-be5b-7ebebd31cccf)

**Comment by dglent** [comment]
In the log file you have some usefull data
`Actual weather status for: Loughton GB 21.9°

Feels like 22 °C

overcast clouds`

I don’t know how you add data in conky. If you read this file every hour for exemple you can exctract the current weather status. If you need more data as forecast it will be easy to add also.
Tell me if this aproach is ok and the needed data.

**Comment by Davetheraveuk** [comment]
@dglent
Hi again and thanks for getting back to me so soon. :)
As requested, below is the data I am looking to place on the background desktop via conky:

Location
Day high
Day low
Feels like
Wind
Cloudiness
Humidity
Visibility
Comfort
Precipitation
Sunrise
Sunset
Daylight
Air quality
UV level


Thanks, Dave.

**Comment by dglent** [comment]
You can add this line:
``` python
logging.info(
                f'EXPORT:{ {key: value if type(value) == str else " ".join(v for v in value) for key, value in self.weatherDataDico.items()}                 }'
            )
```
here just after this line:
https://github.com/dglent/meteo-qt/blob/92ec15380a227fc28ddb0179ab51639cfcbdd4a6/meteo_qt/meteo_qt.py#L2431

and tell me if you can use it and if you want more data

**Comment by Davetheraveuk** [comment]
Hi, I added the line you told me to and now the program will not start at
startup. I try to start the program manually via the start-menu and
nothing. I have since removed the line:

logging.info(
>                 f'EXPORT:{ {key: value if type(value) == str else " ".join(v for v in value) for key, value in self.weatherDataDico.items()}                 }'
>             )
>
>
and the program works fine again.

**Comment by dglent** [comment]
It should work, try the exact following code.
I wanted to know if this way is suitable to import data in Conky.
Then i will format the data (e.g. the time) and i will add more data in this object/dictionary export but firstly test if it is usable like this  

![εικόνα](https://github.com/dglent/meteo-qt/assets/4337756/be4a791d-d77a-4dac-87b0-381cf2f60b09)

``` python
        if not self.toggle_tray_action:
            logging.info(f'Actual weather status for: {self.notification}')
            logging.info(
                f'EXPORT:{ {key: value if type(value) == str else " ".join(v for v in value) for key, value in self.weatherDataDico.items()} }'
            )
```

**Comment by Davetheraveuk** [comment]
Hi,
So I got it working, all works well.
You mentioned if I required any other parameters I should ask. Can
*moonrise* and *moonfall* be added?
Also, can the exported data be in CSV database format? Like:

Parameter,Value
Sunrise,06:48
Sunset,19:03
Feels_Like,10.52 C
...

Dave.

**Comment by dglent** [comment]
I started with this, is the lines format ok for your need ?
```python
logging.info(
    f"\nEXPORT_START\n"
    f"City,{city_label}\n"
    f"Feels like,{self.feels_like_value.text()}\n"
    f"Wind,{self.weatherDataDico['Wind'][4]} {wind_speed} {self.unit_system_wind} {self.weatherDataDico['Wind'][1]}\n"
    f"Cloudiness,{self.weatherDataDico['Clouds']}\n"
    f"EXPORT_END\n"
)
```
Add this code here (it is at the end of the update of the gui panel):
https://github.com/dglent/meteo-qt/blob/92ec15380a227fc28ddb0179ab51639cfcbdd4a6/meteo_qt/meteo_qt.py#L821

![εικόνα](https://github.com/dglent/meteo-qt/assets/4337756/4ecca40d-24b6-40b3-855c-bf42c5a315ad)

**Comment by Davetheraveuk** [comment]
Yes, its fine. Thank you.

**Comment by dglent** [comment]
Try the above commit to tell me if ok
* The air quality an UV are arriving after the update of the GUI so you will see them after the next update of the weather. 
It needs to make some modifications to make some variables global etc, check as it and tell me

**Comment by Davetheraveuk** [comment]
Hi, I cannot seem to find the exported information in a CSV database file
anywhere. Where would the CSV file be saved to? Thanks. David.

**Comment by dglent** [comment]
Ah sorry i didn't precise, the data are logged in the log file which you can find in the folder ~/.config/meteo-qt/meteo-qt.log
You can have a script which extracts (for e.g. every 30 minutes) the data from this file.
You can extract the data and use them directly or create a csv file.
You can use the following awk command:
``` console
[dglent@localhost meteo-qt]$ pwd
/home/dglent/.config/meteo-qt
[dglent@localhost meteo-qt]$ awk '/EXPORT_START/{ f=1;r=""; next }f && /EXPORT_END/{f=0} f{ r=(r=="")? $0: r RS $0 }END{ print r }' meteo-qt.log
City,Paris, FR
Temperature,13.87 °C
Feels like,13.79 °C
Wind,South 7.6  km/h  light breeze
Cloudiness,100% overcast clouds
Humidity,95 %
Visibility,10.0 km
Comfort,Comfortable
Precipitation,rain 1.15  mm 
Sunrise,08:17
Sunset,18:52
Daylight,10:35
Air quality,Moderate
UV,2.08 Low
[dglent@localhost meteo-qt]$ awk '/EXPORT_START/{ f=1;r=""; next }f && /EXPORT_END/{f=0} f{ r=(r=="")? $0: r RS $0 }END{ print r }' meteo-qt.log > meteo_exp.csv
[dglent@localhost meteo-qt]$ ls
meteo_exp.csv  meteo-qt.conf  meteo-qt.log
```

**Comment by Davetheraveuk** [comment]
Hi again, I'm sorry to be a PITA but the CSV file has no data inside it. I
executed the following command:
*awk '/EXPORT_START/{ f=1;r=""; next }f && /EXPORT_END/{f=0} f{ r=(r=="")?
$0: r RS $0 }END{ print r }' meteo-qt.log > meteo_exp.csv*

But the file was empty. I have attached the generated file.
Thanks, Dave.

**Comment by dglent** [comment]
If you run the commande without the "> meteo_exp.csv" you should see the data in the terminal. If not are you sure do have downloaded the latest meteo_qt.py file from github ? Do you find these data in the log file ?

**Comment by Davetheraveuk** [comment]
OK so I downloaded the new meteo_qt.py file from github and yes I see the
data in the log file. Thanks. David.

**Comment by Davetheraveuk** [comment]
Hi @dglent, I have been having problems with the API but I can still see the weather data in the meteo-qt application window. But nothing is displayed within conky. I cleared the meteo-qt log file and restarted the application, below is the information from the log file:

```
2024/09/18 21:25:16 ERROR: Your openweathermap key has no access to onecall api - 2781: meteo_qt
2024/09/18 21:25:16 INFO: Actual weather status for: Loughton GB 15.1°
Feels like 14.18 °C
clear sky - 2431: meteo_qt

```

Even with a new API key the error messages are like what I have posted above.

Thanks, David.
